test/Main.hs

```diff
@@ -25,47 +25,74 @@ tests :: TestTree
+    withFrozenLastQuery $ do
+      void . runQuery $ mkSQL "SELECT 2"
+      getLastQuery >>= \(SomeSQL lastQuery) ->
+        liftIO $ assertEqual "The last query before freeze should be reported" (show sql) (show lastQuery)
+    getLastQuery >>= \(SomeSQL lastQuery) ->
+      liftIO $ assertEqual "The last query before freeze should be reported" (show sql) (show lastQuery)
 
+testConnectionStatsWithNewConnection :: Assertion
+testConnectionStatsWithNewConnection = do
+  dbUrl <- T.pack <$> getEnv "DATABASE_URL"
+  let connectionSource :: ConnectionSource [MonadBase IO, MonadMask]
+      connectionSource = simpleSource $ ConnectionSettings dbUrl Nothing []
+      transactionSettings =
+        defaultTransactionSettings
+          { tsIsolationLevel = ReadCommitted
+          , tsAutoTransaction = False
+          }
+  void . runEff . runErrorNoCallStack @HPQTypesError . runEffectDB connectionSource transactionSettings $ do
+    do
+      void . runQuery $ mkSQL "SELECT 1"
+      void . runQuery $ mkSQL "SELECT 2"
+      connectionStats <- getConnectionStats
+      liftIO $ assertEqual "Incorrect connection stats" (ConnectionStats 2 2 2 0) connectionStats
+    do
+      void . runQuery $ mkSQL "CREATE TABLE some_table (field INT)"
+      void . runQuery $ mkSQL "BEGIN"
+      void . runQuery $ mkSQL "INSERT INTO some_table VALUES (1)"
+      withNewConnection $ do
         connectionStats <- getConnectionStats
-        liftIO $ putStr "Connection stats: " >> print connectionStats
-
-        setTransactionSettings $ defaultTransactionSettings {tsIsolationLevel = ReadCommitted}
-        void . runQuery $ mkSQL "CREATE TABLE some_table (field INT)"
-        void . runQuery $ mkSQL "BEGIN"
-        void . runQuery $ mkSQL "INSERT INTO some_table VALUES (1)"
-        withNewConnection $ do
-          newConnectionStats <- getConnectionStats
-          liftIO $ putStr "New connection stats: " >> print newConnectionStats
-
-          setTransactionSettings $ defaultTransactionSettings {tsIsolationLevel = ReadCommitted}
-          noOfResults <- runQuery $ mkSQL "SELECT * FROM some_table"
-          liftIO $ assertEqual "Results should not be visible yet" 0 noOfResults
-        void . runQuery $ mkSQL "COMMIT"
+        liftIO $ assertEqual "Connection stats should be reset" (ConnectionStats 0 0 0 0) connectionStats
         noOfResults <- runQuery $ mkSQL "SELECT * FROM some_table"
-        liftIO $ assertEqual "Results should be visible" 1 noOfResults
-        void . runQuery $ mkSQL "DROP TABLE some_table"
-  (runEff . runErrorNoCallStack @HPQTypesError $ runEffectDB connectionSource transactionSettings program) >>= print
+        liftIO $ assertEqual "Results should not be visible yet" 0 noOfResults
+      void . runQuery $ mkSQL "COMMIT"
+      noOfResults <- runQuery $ mkSQL "SELECT * FROM some_table"
+      liftIO $ assertEqual "Results should be visible" 1 noOfResults
+      void . runQuery $ mkSQL "DROP TABLE some_table"
```

Comment:
I think we could also have an original test to prove that transaction isolation works between connections.

Reply:
Tests need sorting out anyway (e.g. if there's an error, the table is not removed and subsequent test runs fail), but let's do that in a separate PR.

Reply:
Ok, let's create a ticket once this is merged and improve the tests in a separate PR.